### PR TITLE
more tests for SimpleInterval. 

### PR DESCRIPTION
**Pull request by akiezun**
coverage now at 100% 

fixes https://github.com/broadinstitute/gatk/issues/1405 by moving getSpanningInterval as private to ShuffleJoinReadsWithRefBases

@lbergelson can you review?

### CODE REVIEW

**Comment by lbergelson** [comment]
+1 to increasing test coverage.  

I don't think moving `getSpanningInterval` is the right solution to 1405.  Why not just add the checks and leave it in SimpleInterval?  It's a useful function that will be reinvented in a number of places if it's moved out.

**Comment by droazen** [comment]
Agree with @lbergelson on this one

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java [hunk 1]
```diff
@@ -97,6 +97,7 @@ public SimpleInterval(final String str){
          * So we can either keep this parsing code in the constructor or make a static factory method
          * and make multiple objects. We chose the former.
          */
+        Utils.nonNull(str);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: We should probably add a check for empty string as well.  I don't think  the interval from 1 to max-int on conting "" is something that people would ever want.

**Comment by akiezun** [comment]
If it's useful well move it back. Right now it's used in 1 place.

On Thursday, February 18, 2016, Louis Bergelson <notifications@github.com>
wrote:

> +1 to increasing test coverage.
>
> I don't think moving getSpanningInterval is the right solution to 1405.
> Why not just add the checks and leave it in SimpleInterval? It's a useful
> function that will be reinvented in a number of places if it's moved out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1497#issuecomment-185820758>.
>


-- 
Sent from Gmail Mobile

**Comment by droazen** [comment]
@akiezun What will actually happen is that someone will need that functionality months from now, forget that it already exists (embedded in some random tool), and re-implement it. It should be moved back now before this is allowed to happen.

**Comment by akiezun** [comment]
Fine but this is clearly premature optimization.

How about a class called Intervals or intervalutils for this sort of random
utility ?

On Thursday, February 18, 2016, droazen <notifications@github.com> wrote:

> @akiezun <https://github.com/akiezun> What will actually happen is that
> someone will need that functionality months from now, forget that it
> already exists (embedded in some random tool), and re-implement it. It
> should be moved back now before this is allowed to happen.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1497#issuecomment-185886728>.
>


-- 
Sent from Gmail Mobile

**Comment by droazen** [comment]
`IntervalUtils` already exists -- I'd be fine with moving it there as a compromise. We just need to avoid the classic GATK3 mistake of embedding general utilities within tools.

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: ok

**Comment by akiezun** [comment]
done. moved getSpanningInterval to IntervalUtils and simplified + added tests for it. Moved IntervalUtilsUnitTest to the right package. No empty string in ctor.

back to @lbergelson

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 2]
```diff
@@ -59,14 +59,34 @@
...
+    public static SimpleInterval getSpanningInterval(final List<? extends Locatable> locations) {
+        Utils.nonNull(locations);
+        if (locations.isEmpty()){
+            return null;
+        }
+        final List<String> contigs = locations.stream().map(l -> l.getContig()).distinct().collect(Collectors.toList());
+        if (contigs.size() != 1){
+            throw new IllegalArgumentException("found different contigs from inputs:" + contigs);
+        }
+        final int minStart = locations.stream().mapToInt(l -> l.getStart()).min().getAsInt();
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: I think this doesn't actually fix #1405 yet.

**Comment by lbergelson** [comment]
@akiezun Add that per element null check that valentine wants and then feel free to merge.

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: fixed now by the call to Utils.containsNoNull

**Pull request merged**